dwh/indicator_lib.py

```diff
@@ -82,14 +82,48 @@ def calc_TX_PVLS(patient_agg_obs: pandas.DataFrame, VL_code: str,
   if end_date_str:
     end_date = datetime.strptime(end_date_str, '%Y-%m-%d')
   temp_df = patient_agg_obs[(patient_agg_obs['code'] == VL_code)].copy()
+  # separate value from date
+  temp_df[['vl_max_date', 'latest_vl_value']] = temp_df.max_date_value.str.split('_SeP_', expand=True)
+  temp_df['latest_vl_value'] = temp_df['latest_vl_value'].astype(float)
   # Note the above copy is used to avoid setting a new column on a slice next:
-  temp_df['sup_VL'] = (temp_df['max_value'] < failure_threshold)
+  temp_df['sup_VL'] = (temp_df['latest_vl_value'] < failure_threshold)
   temp_df['buckets'] = temp_df.apply(
       lambda x: _agg_buckets(x.birthDate, x.gender, datetime.today()), axis=1)
   temp_df_exp = temp_df.explode('buckets')
   temp_df_exp = temp_df_exp.groupby(['sup_VL', 'buckets'], as_index=False)\
       .count()[['sup_VL', 'buckets', 'patientId']]\
-      .rename(columns={'patientId': 'count'})
+      .rename(columns={'patientId': 'sup_VL_count'})
   # calculate ratio
-  temp_df_exp['ratio'] = temp_df_exp['count']/temp_df['sup_VL'].count()
+  temp_df_exp['sup_VL_ratio'] = temp_df_exp['sup_VL_count']/temp_df['sup_VL'].count()
```

Comment:
While I personally like the `[]` notation on data-frames but I think its use is discouraged in cases like this. I think you should get a `SettingWithCopyWarning` when doing this, can you please try and confirm?

The issue is described [here](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#indexing-view-versus-copy); the gist of it is that when we slice a DF it is not guaranteed whether it would be a "view" or a "copy", hence setting a new column on that slice is not well defined. That's the reason that you see the explicit `copy()` above on line 84.

So after confirming that you will actually get that warning, I think the fix is either to make an explicit copy after your projection on line 94 or using `.loc` notation (I should have made this comment on your previous PR when ratio was added).

Reply:
Another option is to use [`assign()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.assign.html).

Reply:
noted with thanks, this is guideline is new to me

Reply:
I did not see any warnings, but just to be thorough I have used `.loc`